File: 3682-count-the-number-of-arrays-with-k-matching-adjacent-elements/count-the-number-of-arrays-with-k-matching-adjacent-elements.rb
MOD = 1_000_000_007
MAXN = 100001
$fact = Array.new(MAXN, 0)

# Giai thừa với memo
def factorial(n)
  return $fact[n] if $fact[n] != 0
  return $fact[0] = 1 if n == 0
  $fact[n] = factorial(n - 1) * n % MOD
end

# Lũy thừa nhanh (a^b % mod)
def power(a, b)
  result = 1
  base = a % MOD
  while b > 0
    result = result * base % MOD if b.odd?
    base = base * base % MOD
    b >>= 1
  end
  result
end

# Nghịch đảo modulo theo Fermat
def mod_inv(a)
  power(a, MOD - 2)
end

# C(n, k)
def comb(n, k)
  return 0 if k < 0 || k > n
  factorial(n) * mod_inv(factorial(k)) % MOD * mod_inv(factorial(n - k)) % MOD
end

# Hàm chính
def count_good_arrays(n, m, k)
  diff = n - 1 - k
  m * power(m - 1, diff) % MOD * comb(n - 1, diff) % MOD
end

File: 3682-count-the-number-of-arrays-with-k-matching-adjacent-elements/count-the-number-of-arrays-with-k-matching-adjacent-elements.go
const mod = 1_000_000_007
const maxN = 100001

var fact [maxN]int

// Khởi tạo giai thừa nếu chưa có
func getFactorial(n int) int {
	if fact[n] != 0 {
		return fact[n]
	}
	if n == 0 {
		fact[0] = 1
		return 1
	}
	fact[n] = getFactorial(n-1) * n % mod
	return fact[n]
}

// Lũy thừa nhanh (a^b % mod)
func power(a, b int) int {
	res := 1
	base := a % mod
	for b > 0 {
		if b%2 == 1 {
			res = res * base % mod
		}
		base = base * base % mod
		b /= 2
	}
	return res
}

// Tính nghịch đảo modulo bằng Fermat: a^(mod-2) % mod
func modInverse(a int) int {
	return power(a, mod-2)
}

// C(n, k) = n! / (k! * (n-k)!)
func comb(n, k int) int {
	if k < 0 || k > n {
		return 0
	}
	return getFactorial(n) * modInverse(getFactorial(k)) % mod * modInverse(getFactorial(n-k)) % mod
}

// Hàm chính
func countGoodArrays(n, m, k int) int {
	// Số phần tử khác nhau: (n-1) - k
	diff := n - 1 - k
	return m * power(m-1, diff) % mod * comb(n-1, diff) % mod
}